_posts/2019-04-10-imperatywnie-vs-funkcyjnie.md

```diff
@@ -0,0 +1,118 @@
+
+### Dajmy przykład, w którym dodajesz dwie liczby:
+* W podejściu imperatywnym dodajemy liczby i tworzymy wynik. Skupiamy się na tym co chcemy zrobić.
+* W podejściu funkcyjnym dodane liczby to jest wynik. (drobna, a jednak znaczna różnica o czym się zaraz przekonamy)
+
+### Imperatywny przykład
+Chcemy tylko nieparzyste liczby: 
+
+```kotlin
+val numbers = listOf(1, 2, 3, 4, 5)
+val odds = ArrayList<Int>()
+
+for (index in 0..numbers.lastIndex) {
+    val item = numbers[index]
+    if (item % 2 != 0) odds.add(item)
+}
+```
+Powyższy kod jest zbudowany z wyrażeń. Coś bierzemy, coś dodajemy, coś zmieniamy, coś warunkujemy. Mutujemy zmienne przez co kod nie jest bezpieczny. W obecnych funkcyjnych czasach mimo wszystko rzadko spotyka się powyższy kod i dobrze, bo jest on kompletnie nieczytelny. Skupiamy się tutaj na tym co chcemy zrobić. Jest to po prostu seria mutacji oddzielonych od siebie warunkami.
+
+### Funkcyjny przykład
+```kotlin
+val numbers = listOf(1, 2, 3, 4, 5)
+val odds = numbers.filter { it % 2 != 0 }
+```
+W końcu coś z czym większość z nas spotka się na codzień, czyli funkcyjny kod. Proste i szybkie nie potrzeba nawet komentarza. Skupiamy się na tym co chcemy osiągnać. 
+
+## Czy da się napisać program bez efektu ubocznego? 
+Krótka odpowiedź. Nie da się. Chodzi nam bardziej o to, żeby nie mieć obserwowalnych efektów ubocznych. Co to znaczy? Wyjaśnimy sobie już za chwilkę. Zazwyczaj jak piszemy apkę to mamy widoczny efekt - wynik. Zapisaliśmy coś do bazy danych, wysłaliśmy coś po HTTPie, wrzuciliśmy jakiś event na kolejkę, wygenerowaliśmy raport i tak dalej. Integrujemy się ze światem zewnętrznym. W programowaniu funkcyjnym chodzi o odłożenie efektów ubcznych do czasu  wykonania obliczeń, a nie podczas ich.
+
+```kotlin
+fun multiply(val a: Int, val b: Int) = a * b
+fun divide(a: Int, b: Int) = a / b
+ 
+multiply(123456789, 123456789) // -1757895751
+divive(123456789, 0) // ArithmeticException: / by zero
+```
+Powyższy program nie jest funkcyjny dlatego, że mogą w nim wystąpić efekty uboczne. Jeśli przekręcimy Inta to dostaniemy ujemną wartość. W drugim przypadku dostaniemy wyjątek czego również się nie spodziewaliśmy. Nasze początkowe założenie, że funkcja pomnoży/podzieli wynik jest błędna. Oczywiście rzadko kiedy sytuacja jest, aż tak trywialna, ale można sobie wyobrazić, że to wszystko jest jakimś zapytaniem do bazy, albo inną operacją. Chodzi o to, aby nasze funkcje były deterministyczne, czyli wynik zawsze jest taki sam dla podanych argumentów. Jeśli są jakieś efekty uboczne to wiemy, jakie i odpowiednio na nie reagujemy. 
+
+```kotlin
+fun multiply(a: Int, b: Int) = a * b.toFloat()
+fun divide(a: Int, b: Int) = a / b.toFloat()
+ 
+multiply(123456789, 123456789) // 1.52415794E16
+divide(123456789, 0) // Infinity
+
+result.isInfinite().maybe { print("You should learn the basics of math, probably.") }
+```
+
+W tym momencie dla każdej możliwej wartości będzie ten sam wynik. Również w przypadku poprzedniego wyjątku teraz 
```

Comment:
Ej, miałeś się skupić na programowaniu funkcyjnym a nie na bajerach języka i bibliotek. Clue jest takie, że dostajesz Infinity a nie jakiś wyjątek

Reply:
Trochę usunąłem, ale chyba warto napisać czym jest `maybe {  }`? 
Tym bardziej, że jest z zewnętrznej biblioteki.

Reply:
Pytanie czy to maybe jest Ci do czegokolwiek potrzebne? IMO tak naprawdę wystarczy informacja, że dzielenie przez zero zwróci obiekt a nie rzuci wyjątkiem - nie ma efektów ubocznych.

Reply:
trochę zmieniłem